### PR TITLE
[Lake] Add lake connector

### PR DESCRIPTION
**Pull request by wyb**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Problem Summary(Required) ：

1. Add new TLakeScanNode for lake table.
2. Add lake connector for lake tablet scan.

Parallel scan within a single tablet is not supported.

### CODE REVIEW

File: be/src/connector/CMakeLists.txt [hunk 1]
```diff
@@ -8,5 +8,6 @@ add_library(Connector STATIC
     hive_connector.cpp
     es_connector.cpp
     jdbc_connector.cpp
+    lake_connector.cpp
```
Comment (dirtysalt) [inline be/src/connector/CMakeLists.txt:11]: deserves a better & more specific name.

File: be/src/connector/lake_connector.cpp [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+
+DataSourceProviderPtr LakeConnector::create_data_source_provider(vectorized::ConnectorScanNode* scan_node,
+                                                                 const TPlanNode& plan_node) const {
+    return std::make_unique<LakeDataSourceProvider>(scan_node, plan_node);
+}
+
+// ================================
+
+LakeDataSourceProvider::LakeDataSourceProvider(vectorized::ConnectorScanNode* scan_node, const TPlanNode& plan_node)
+        : _scan_node(scan_node), _t_olap_scan_node(plan_node.olap_scan_node) {}
```
Comment (dirtysalt) [inline be/src/connector/lake_connector.cpp:26]: why `_t_`?

**Review by dirtysalt — DISMISSED** [review]

File: be/src/connector/lake_connector.cpp (hunk 2 above)
Comment (wuleistarrocks) [reply be/src/connector/lake_connector.cpp:26]: > why `_t_`?

't' means thrift?

**Review by wuleistarrocks — DISMISSED** [review]

File: be/src/connector/lake_connector.cpp (hunk 2 above)
Comment (wyb) [reply be/src/connector/lake_connector.cpp:26]: > > why `_t_`?
> 
> 't' means thrift?

yes

File: be/src/connector/CMakeLists.txt (hunk 1 above)
Comment (wyb) [reply be/src/connector/CMakeLists.txt:11]: It is for StarRocks lake table.

File: be/src/connector/lake_connector.h [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    LakeDataSourceProvider(vectorized::ConnectorScanNode* scan_node, const TPlanNode& plan_node);
+    DataSourcePtr create_data_source(const TScanRange& scan_range) override;
+
+protected:
+    vectorized::ConnectorScanNode* _scan_node;
+    const TOlapScanNode _t_olap_scan_node;
+};
+
+// TODO: support parallel scan within a single tablet
+class LakeDataSource final : public DataSource {
```
Comment (sduzh) [inline be/src/connector/lake_connector.h:37]: Can we move the class definition to cpp file to reduce compile-time dependency.

File: gensrc/thrift/PlanNodes.thrift [hunk 4]
```diff
@@ -318,6 +318,7 @@ struct TOlapScanNode {
   23: optional map<i32, i32> dict_string_id_to_int_ids
   // which columns only be used to filter data in the stage of scan data
   24: optional list<string> unused_output_column_name
+  25: optional bool is_lake_table
```
Comment (sduzh) [inline gensrc/thrift/PlanNodes.thrift:321]: Is it appropriate to add a new `TLakeScanNode`, instead of reuse the `TOlapScanNode`?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/connector/lake_connector.h (hunk 3 above)
Comment (wyb) [reply be/src/connector/lake_connector.h]: done

**Comment by wyb** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/connector/lake_connector.cpp [hunk 5]
```diff
@@ -0,0 +1,598 @@
...
+    void close(RuntimeState* state) override;
+    Status get_next(RuntimeState* state, vectorized::ChunkPtr* chunk) override;
+
+    int64_t raw_rows_read() const override { return _raw_rows_read; }
+    int64_t num_rows_read() const override { return _num_rows_read; }
+    int64_t num_bytes_read() const override { return _bytes_read; }
+
+    // TODO:
+    // Return last bytes of Scan or Exchange Data, then reset it
+    // int64_t last_spent_cpu_time_ns() override;
```
Comment (mofeiatwork) [inline be/src/connector/lake_connector.cpp:41]: Please refer this PR to implement the `get_cpu_time` interface: https://github.com/StarRocks/starrocks/pull/8088

File: be/src/connector/lake_connector.cpp [hunk 6]
```diff
@@ -0,0 +1,598 @@
...
+
+    RETURN_IF_ERROR(build_scan_range(_runtime_state));
+
+    RETURN_IF_ERROR(init_tablet_reader(_runtime_state));
+    return Status::OK();
+}
+
+void LakeDataSource::close(RuntimeState* state) {
+    update_counter();
+    _prj_iter->close();
```
Comment (mofeiatwork) [inline be/src/connector/lake_connector.cpp:209]: Please check these pointers before close.

If open failed, these pointers would be `nullptr` and cause a crash here.

File: gensrc/thrift/PlanNodes.thrift (hunk 4 above)
Comment (wyb) [reply gensrc/thrift/PlanNodes.thrift]: done

File: be/src/connector/lake_connector.cpp [hunk 7]
```diff
@@ -0,0 +1,598 @@
...
+void LakeDataSource::decide_chunk_size() {
+    bool has_huge_length_type = std::any_of(_query_slots.begin(), _query_slots.end(),
+                                            [](auto& slot) { return slot->type().is_huge_type(); });
+    if (_read_limit != -1 && _read_limit < _runtime_state->chunk_size()) {
+        // Improve for select * from table limit x, x is small
+        _params.chunk_size = _read_limit;
+    } else {
+        _params.chunk_size = _runtime_state->chunk_size();
+    }
+    if (has_huge_length_type) {
```
Comment (mofeiatwork) [inline be/src/connector/lake_connector.cpp:328]: This check is unnecessary, current default `chunk_size` has been changed to `4096`.

File: be/src/connector/lake_connector.cpp [hunk 8]
```diff
@@ -0,0 +1,598 @@
...
+        _params.chunk_size = std::min(_params.chunk_size, CHUNK_SIZE_FOR_HUGE_TYPE);
+    }
+}
+
+Status LakeDataSource::init_reader_params(const std::vector<OlapScanRange*>& key_ranges,
+                                          const std::vector<uint32_t>& scanner_columns,
+                                          std::vector<uint32_t>& reader_columns) {
+    const TOlapScanNode& thrift_olap_scan_node = _provider->_t_olap_scan_node;
```
Comment (mofeiatwork) [inline be/src/connector/lake_connector.cpp]: Why use the `OlapScanNode` but not a new ExecNode?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/connector/lake_connector.cpp [hunk 9]
```diff
@@ -0,0 +1,598 @@
...
+    bool skip_aggregation = thrift_olap_scan_node.is_preaggregation;
```
Comment (mofeiatwork) [inline be/src/connector/lake_connector.cpp]: Why not share the code with `OlapChunkSize`?

File: be/src/connector/lake_connector.cpp (hunk 8 above)
Comment (wyb) [reply be/src/connector/lake_connector.cpp]: done

File: be/src/connector/lake_connector.cpp (hunk 6 above)
Comment (wyb) [reply be/src/connector/lake_connector.cpp]: done

File: be/src/connector/lake_connector.cpp (hunk 7 above)
Comment (wyb) [reply be/src/connector/lake_connector.cpp]: removed

File: be/src/connector/lake_connector.cpp (hunk 9 above)
Comment (wyb) [reply be/src/connector/lake_connector.cpp]: I will refactor it later.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by wuleistarrocks — DISMISSED** [review]

**Review by murphyatwork — DISMISSED** [review]

**Comment by mofeiatwork** [comment]
Why not implement it another `ScanOperator` and `ChunkSource` so that you could reuse most of existed code ?

**Review by sduzh — APPROVED** [review]

**Comment by wyb** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by dirtysalt — APPROVED** [review]

**Pull request merged**